src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt

```diff
@@ -275,137 +276,159 @@ class Peer(
         )
     }
 
-    fun connect() {
-        if (connectionState.value is Connection.CLOSED) {
+    data class ConnectionJob(val job: Job, val socket: TcpSocket) {
+        fun cancel() {
+            job.cancel()
+            socket.close()
+        }
+    }
+
+    private var connectionJob: ConnectionJob? = null
+
+    suspend fun connect(connectTimeout: Duration, handshakeTimeout: Duration): Boolean {
+        return if (connectionState.value is Connection.CLOSED) {
+            // Clean up previous connection state: we do this here to ensure that it is handled before the Connected event for the new connection.
+            // That means we're not sending this event if we don't reconnect. It's ok, since that has the same effect as not detecting a disconnection and closing the app.
+            input.send(Disconnected)
             _connectionState.value = Connection.ESTABLISHING
-            establishConnection()
+
+            val connectionId = currentTimestampMillis()
+            val logger = MDCLogger(nodeParams.loggerFactory.newLogger(this::class), staticMdc = mapOf("remoteNodeId" to remoteNodeId, "connectionId" to connectionId))
+            logger.info { "connecting to ${walletParams.trampolineNode.host}" }
+            val socket = openSocket(connectTimeout) ?: return false
```

Comment:
If `openSocket` returns null and `connectionState` is NOT set to `CLOSED`, the `connect` method would return and the connection state would remain `CONNECTING`, i.e. we'd be deadlocked.

Looking at `openSocket` it seems that the only scenario where it can return null will be `caught` first -- and that catch does update the connection state to `CLOSED`. So we are probably OK and the deadlock scenario ^ is not possible. Can you confirm?

Reply:
Yes indeed: since `socketBuilder.connect` always returns a non-null `TcpSocket`, the only case where we can end up with `null` is if we went through the `catch` handler that resets the `_connectionState` to `CLOSED`, so we should be fine!